### DISCUSSION TITLE
Prettify

### DISCUSSION THREAD

**Discussion by orefalo** [lmorg/murex#721] [Ideas]
Just opening this topic to discuss prettiness and look and feel.

**Comment by orefalo** [comment]
This is what `kill` completion looks like in the ocean

<img width="1172" alt="CleanShot 2023-09-06 at 10 08 58" src="https://github.com/lmorg/murex/assets/52085/526fd270-1804-4036-9a73-4918ad05e400">

**Reply by lmorg** [reply]
I can mimic this in Murex. It shouldn't be a big change to readline

**Reply by orefalo** [reply]
far from urgent,
just collecting what I find in this thread

shell users are console lovers ;-)

**Comment by orefalo** [comment]
Similar topic, related to https://github.com/lmorg/murex/issues/683

this what the command line looks like in other oceans

<img width="444" alt="CleanShot 2023-09-06 at 10 11 29" src="https://github.com/lmorg/murex/assets/52085/c4559f48-5ce5-4427-8178-5a52301b67f7">


Is it more than theming?

**Reply by lmorg** [reply]
I want to rewrite the syntax highlighter. Currently it is its own parser (designed for performance instead of accuracy) but that parser is quickly becoming unmaintainable. So I've been mulling over the idea of incorporating the syntax highlighting into the main parser. Unfortunately that's not quite as easy as it should be because the main parser fails early on error (again for performance releases -- there's no point parsing an entire script if there is an error on the first line that prevents the rest of the script from executing). Additionally I'd still need a parser for the syntax completion.

Maybe the solution here it to write a completely new parser for highlighting and completion?

As for theming, I might expose that as user configuration in the next release after v5.0

**Reply by orefalo** [reply]
can't wait to see 5.0 released.
  it's the trigger point for `brew` and `starship` enablement

**Reply by lmorg** [reply]
I've done some work on rewriting the syntax highlighter, with some emphasis on exposing those values as user customisable fields. Unfortunately I've hit some show stopper bugs that will require rewriting the code...again. But actually I don't mind since the solution I was originally writing felt more and more hacky the more code I wrote. The other side effect is I'll also have to rewrite the spell checking code too. Needless to say, this is proving a bigger piece of work than I'd hoped.

**Comment by orefalo** [comment]
Another one - autocomplete seems "smarter" in the ocean

<img width="583" alt="CleanShot 2023-09-10 at 17 40 36@2x" src="https://github.com/lmorg/murex/assets/52085/56f58e49-aa64-4a86-a699-2241b14d3f9b">

in the above, the cmd line stop at `chsh`, by default it shows a possible history match in a different color
pressing TAB is on spot with the chsh functions.

**Reply by lmorg** [reply]
I never added inline history completion because, personally, I don't like that feature much. My issue with it is that it isn't deterministic in that much of the stuff I do in the command line is down to muscle memory and since history changes, it means you cannot develop muscle memory around history. Plus I think it makes the command line a lot noisier.

That all said, I'm not against adding a feature I don't personally like :) 

I will share a couple of features that are implemented and a little more deterministic which might compensate for the lack of inline history matching...and if not I'll have a think about how I could implement inline history matching.

1. `ctrl`+`r`, this can be used at any point and will bring up the fuzzy finder for history matches based on your command line. You can type multiple partial words, globs, and even regexp expressions so search through it. Though this feature is slightly better suited for finding that weird command line you typed months ago rather than repeating common command lines.

2. history macros, which are basically bookmarks in your history file. Bookmarks are added by appending `#bookmarkname` (no spaces between `#` and the bookmark name) to the end of your command line eg

<img width="960" alt="image" src="https://github.com/lmorg/murex/assets/12153701/2cce5888-2db2-41d8-afba-dd24202114b8">

<img width="960" alt="image" src="https://github.com/lmorg/murex/assets/12153701/795eff51-745a-45ad-b659-03822575618a">

You can also add variables to your macros with `^$varname`:
<img width="960" alt="image" src="https://github.com/lmorg/murex/assets/12153701/3ad36cf5-a0a8-4670-b6c4-76c502618ae5">
this is handy for those occasions you need to rerun the same command line repeatedly but with slightly different values, but it's not a task that really benefits from an alias or function (eg it might already be single command so creating a function around it doesn't simplify it's usage. or it might be a task that you're working on for a week and then nevermore).

**Reply by lmorg** [reply]
@orefalo what's your thoughts on the above? Does that scratch your itch or does it not solve your use case?

**Reply by orefalo** [reply]
sorry for the late reply, I also took a break ;-)

I should have clarified, the screenshot above is not inline history.
it's contextual autocomplete: I just typed `chsh <TAB>` in both shells

<img width="596" alt="CleanShot 2023-09-20 at 12 53 00" src="https://github.com/lmorg/murex/assets/52085/6bb15f3a-0b5e-4e2b-9b7b-54cec0fa658e">

<img width="576" alt="CleanShot 2023-09-20 at 12 53 58" src="https://github.com/lmorg/murex/assets/52085/c68f53a2-73f0-4267-9a3f-1e17484d68c1">

Now agreed, fish also shows inline history along with it (which I personally find useful so I immediately know if UP will match the command I am looking for).
But my point is getting the **user list**, very useful.

the comment is not limited to chsh, I am just pointing to a piece of work where murex probably needs to be smarted about context sensitive parameters.

PS: I didn't know about bookmarks and variables..to be documented I believe ;-)

**Reply by lmorg** [reply]
Thanks for the clarification. This is just custom autocompletes. I've written a few already but there's so many CLI tools out there that it's impossible to write completions for everything. So the ones that just have flags without any context are the default autocompletes as parsed from the man pages.

Unfortunately the only solution to context sensitive autocompletes is to write custom autocompletes. And like with Bash, Fish and other shells, that needs to be a community effort.

> sorry for the late reply, I also took a break ;-)

completely understandable. I needed a few days away from code myself :)

> PS: I didn't know about bookmarks and variables..to be documented I believe ;-)

Yeah. I'll need to write the docs on those though because there are quite a few features there any caveats specific to them (eg they only work in the interactive shell, and are parsed differently too).
